src/Components/Components/test/Auth/AuthenticationStateProviderTest.cs

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Security.Claims;
+using System.Security.Principal;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.RenderTree;
+using Microsoft.AspNetCore.Components.Test.Helpers;
+using Microsoft.Extensions.DependencyInjection;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Components
+{
+    public class AuthenticationStateProviderTest
+    {
+        [Fact]
+        public void RequiresRegisteredService()
+        {
+            // Arrange
+            var renderer = new TestRenderer();
+            var component = new AutoRenderFragmentComponent(builder =>
+            {
+                builder.OpenComponent<AuthenticationStateProvider>(0);
+                builder.CloseComponent();
+            });
+
+            // Act/Assert
+            renderer.AssignRootComponentId(component);
+            var ex = Assert.Throws<InvalidOperationException>(() => component.TriggerRender());
+            Assert.Contains($"There is no registered service of type '{typeof(IAuthenticationStateProvider).FullName}'.", ex.Message);
+        }
+
+        [Fact]
+        public void SuppliesSynchronouslyAvailableAuthStateToChildContent()
+        {
+            // Arrange: Service
+            var services = new ServiceCollection();
+            var authStateProvider = new TestAuthStateProvider()
+            {
+                CurrentAuthStateTask = Task.FromResult<IAuthenticationState>(new TestAuthState("Bert"))
+            };
+            services.AddSingleton<IAuthenticationStateProvider>(authStateProvider);
+
+            // Arrange: Renderer and component
+            var renderer = new TestRenderer(services.BuildServiceProvider());
+            var component = new UseAuthenticationStateProviderComponent();
+
+            // Act
+            renderer.AssignRootComponentId(component);
+            component.TriggerRender();
+
+            // Assert
+            var batch = renderer.Batches.Single();
+            var receiveAuthStateId = batch.GetComponentFrames<ReceiveAuthStateComponent>().Single().ComponentId;
+            var receiveAuthStateDiff = batch.DiffsByComponentId[receiveAuthStateId].Single();
+            Assert.Collection(receiveAuthStateDiff.Edits, edit =>
+            {
+                Assert.Equal(RenderTreeEditType.PrependFrame, edit.Type);
+                AssertFrame.Text(
+                    batch.ReferenceFrames[edit.ReferenceFrameIndex],
+                    "Authenticated: True; Name: Bert; Pending: False; Renders: 1");
+            });
+        }
```

Comment:
As we build more of these type of components, would it make sense for us to build more of a framework for testing these kinds of details. 


Nice to have for sure, but there's a maintenance cost of writing tests that are so coupled to the low level implementation. I'm worried about things like trying to make changes to the render tree or compiler and getting nuked from orbit by our own tests.

Reply:
100% yes. This has been one of the "big rocks" since the beginning: https://github.com/aspnet/AspNetCore/issues/5458

Each time I write these kinds of unit tests, I'm doing so with a view to learn what patterns turn out to be useful. My hope is that eventually, we extract a component unit testing library from our real-world experience of testing our own components.

It's pretty cool to have the chance to design the unit testing system through this approach instead of just abstractly in our heads for a change.

> tests that are so coupled to the low level implementation

I share that concern somewhat. But I'm also really glad that they are coupled to details like "exactly how many renders occurred, and in what order" because it forces us to be aware of certain perf characteristics and would blow up if we regressed it. For "regression", I include things like "taking stuff that was synchronous and making it asynchronous".

Reply:
BTW in case you were wondering, I'd imagine the unit testing mechanism we eventually ship would make your tests look vastly simpler than these, because it would include something like CSS-style selectors for traversing the rendered document (represented as a whole, not separate trees for each component). You wouldn't be coding in terms of render tree frames.

In most cases you'd probably write assertions in terms of the final document shape, not in terms of diffs like we do here. But it is releveant to observe individual diffs if you're being really precise.

Reply:
Agreed on all of these points.